File: asherah-dotnet-shared/src/AppEncryption/Kms/AwsKeyManagementServiceImpl.cs
using System;
using System.Collections.Generic;
using Amazon.Runtime;
using GoDaddy.Asherah.Crypto.Keys;
using Microsoft.Extensions.Logging;

namespace GoDaddy.Asherah.AppEncryption.Kms;

public class AwsKeyManagementServiceImpl : KeyManagementService
{
    internal AwsKeyManagementServiceImpl(Builder builder)
    {
        RegionMap = builder.RegionMap;
        PreferredRegion = builder.PreferredRegion;
        Credentials = builder.Credentials;
        Logger = builder.Logger;
    }

    internal Dictionary<string, string> RegionMap { get; }
    internal string PreferredRegion { get; }
    internal AWSCredentials? Credentials { get; }
    internal ILogger? Logger { get; }

    public static Builder NewBuilder(Dictionary<string, string> regionToArnDictionary, string region) =>
        new(regionToArnDictionary, region);

    public override byte[] EncryptKey(CryptoKey key) =>
        throw new NotSupportedException("AwsKeyManagementServiceImpl is configuration-only when using native core");

    public override CryptoKey DecryptKey(byte[] keyCipherText, DateTimeOffset keyCreated, bool revoked) =>
        throw new NotSupportedException("AwsKeyManagementServiceImpl is configuration-only when using native core");

    public interface IBuildStep
    {
        IBuildStep WithCredentials(AWSCredentials credentials);
        IBuildStep WithLogger(ILogger logger);
        AwsKeyManagementServiceImpl Build();
    }

    public class Builder : IBuildStep
    {
        internal Dictionary<string, string> RegionMap { get; }
        internal string PreferredRegion { get; }
        internal AWSCredentials? Credentials { get; private set; }
        internal ILogger? Logger { get; private set; }

        public Builder(Dictionary<string, string> regionToArnDictionary, string region)
        {
            RegionMap = new Dictionary<string, string>(regionToArnDictionary);
            PreferredRegion = region;
        }

        public IBuildStep WithCredentials(AWSCredentials credentials)
        {
            Credentials = credentials;
            return this;
        }

        public IBuildStep WithLogger(ILogger logger)
        {
            Logger = logger;
            return this;
        }

        public AwsKeyManagementServiceImpl Build() => new(this);
    }
}

File: asherah-java/java/src/main/java/com/godaddy/asherah/appencryption/kms/AwsKeyManagementServiceImpl.java
package com.godaddy.asherah.appencryption.kms;

import com.godaddy.asherah.jni.AsherahConfig;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/**
 * AWS KMS adapter. Maps to kms="aws" in the native config.
 * Compatible with the canonical godaddy/asherah AwsKeyManagementServiceImpl.
 */
public class AwsKeyManagementServiceImpl implements KeyManagementService {

    private final Map<String, String> regionToArnMap;
    private final String preferredRegion;

    private AwsKeyManagementServiceImpl(final Builder builder) {
        this.regionToArnMap = new LinkedHashMap<>(builder.regionToArnMap);
        this.preferredRegion = builder.preferredRegion;
    }

    @Override
    public void applyConfig(final AsherahConfig.Builder builder) {
        builder.kms("aws");
        builder.regionMap(regionToArnMap);
        builder.preferredRegion(preferredRegion);
    }

    public static Builder newBuilder(final Map<String, String> regionToArnMap, final String preferredRegion) {
        return new Builder(regionToArnMap, preferredRegion);
    }

    public static final class Builder {
        private final Map<String, String> regionToArnMap;
        private final String preferredRegion;

        private Builder(final Map<String, String> regionToArnMap, final String preferredRegion) {
            Objects.requireNonNull(regionToArnMap, "regionToArnMap");
            Objects.requireNonNull(preferredRegion, "preferredRegion");
            this.regionToArnMap = new LinkedHashMap<>(regionToArnMap);
            this.preferredRegion = preferredRegion;
        }

        public AwsKeyManagementServiceImpl build() {
            return new AwsKeyManagementServiceImpl(this);
        }
    }
}
